twenty-first/src/shared_math/rescue_prime_regular.rs

```diff
@@ -0,0 +1,1479 @@
+    fn xlix_round(sponge: &mut RescuePrimeRegularState, round_index: usize) {
+        assert!(
+            round_index < NUM_ROUNDS,
+            "Cannot apply {}th round; only have {} in total.",
+            round_index,
+            NUM_ROUNDS
+        );
+
+        // S-box
+        // for i in 0..STATE_SIZE {
+        //     self.state[i] = self.state[i].mod_pow_u64(ALPHA);
+        // }
+        //
+        sponge.state = Self::batch_mod_pow_alpha(sponge.state);
+
+        // MDS matrix
+        let mut v: [BFieldElement; STATE_SIZE] = [BFieldElement::from(0u64); STATE_SIZE];
+        for i in 0..STATE_SIZE {
+            for j in 0..STATE_SIZE {
+                v[i] += BFieldElement::from(MDS[i * STATE_SIZE + j]) * sponge.state[j];
+            }
+        }
+        sponge.state = v;
+
+        // round constants A
+        for i in 0..STATE_SIZE {
+            sponge.state[i] +=
+                BFieldElement::from(ROUND_CONSTANTS[round_index * STATE_SIZE * 2 + i]);
+        }
+
+        // Inverse S-box
+        // for i in 0..STATE_SIZE {
+        //     self.state[i] = self.state[i].mod_pow_u64(ALPHA_INV);
+        // }
+        //
+        // self.state = Self::batch_mod_pow(self.state, ALPHA_INV);
+        sponge.state = Self::batch_mod_pow_alpha_inv(sponge.state);
+
+        // MDS matrix
+        for i in 0..STATE_SIZE {
+            v[i] = BFieldElement::zero();
+            for j in 0..STATE_SIZE {
+                v[i] += BFieldElement::from(MDS[i * STATE_SIZE + j]) * sponge.state[j];
+            }
+        }
+        sponge.state = v;
+
+        // round constants B
+        for i in 0..STATE_SIZE {
+            sponge.state[i] +=
+                BFieldElement::from(ROUND_CONSTANTS[round_index * STATE_SIZE * 2 + STATE_SIZE + i]);
+        }
+    }
+
+    /// xlix
+    /// XLIX is the permutation defined by Rescue-Prime. This
+    /// function applies XLIX to the state of a sponge.
+    fn xlix(sponge: &mut RescuePrimeRegularState) {
+        for i in 0..NUM_ROUNDS {
+            Self::xlix_round(sponge, i);
+        }
+    }
+
+    /// hash_10
+    /// Hash 10 elements, or two digests. There is no padding because
+    /// the input length is fixed.
+    pub fn hash_10(input: &[BFieldElement; 10]) -> [BFieldElement; 5] {
+        let mut sponge = RescuePrimeRegularState::new();
+
+        // absorb once
+        sponge.state[..10].copy_from_slice(input);
+
+        // apply domain separation for fixed-length input
+        sponge.state[10] = BFieldElement::one();
+
+        // apply xlix
+        Self::xlix(&mut sponge);
+
+        // squeeze once
+        sponge.state[..5].try_into().unwrap()
+    }
+
+    /// hash_varlen
+    /// Hash an arbitrary number of field elements. Takes care of
+    /// padding.
+    pub fn hash_varlen(input: &[BFieldElement]) -> [BFieldElement; 5] {
+        let mut sponge = RescuePrimeRegularState::new();
+
+        // pad input
+        let mut padded_input = input.to_vec();
```

Comment:
There is a benchmark for `rescue_prime_regular` which also runs a benchmark for the `varlen` variant, so it is possible to answer this question. However, re-writing the code to avoid the extra clone seems rather tricky. Care to suggest something?

Reply:
I did something quirky once for the `RescuePrimeXlix` struct that apparently conflicted with the way we intend to do padding, so the version that copies less hasn't been in production for three months. I'll try and dig up an old diff. For now, let's assume that this is efficient enough and we might have room to improve.